Consolidate custom post type and meta registrations into Post_Types class

obenland opened the pull request:
This has been on my mental list for a while. Mainly because we register post meta in two places. After that, because the Activitypub class is a junk drawer class and this is something that we can pull out and it makes sense on its own.

## Proposed changes:

This PR consolidates all custom post type and post meta registrations into a single, well-organized `Post_Types` class to improve code maintainability and organization.

* Created new `Post_Types` class with dedicated methods for each post type registration
* Each method handles both post type registration and its related meta fields together
* Moved registrations from scattered locations (`Activitypub::register_post_types()` and `Blocks::register_postmeta()`) into semantic functions
* Maintained user meta registrations in original location as requested
* Updated test to use new class methods
* Removed unused imports after consolidation
* Updated code coverage annotation job to account for multiple warnings

### Other information:

- [x] Have you written new tests for your changes, if applicable? (Updated existing test)

## Testing instructions:

1. Ensure all existing functionality works as expected
2. Run `npm run env-test` to verify all 895 tests pass
3. Check that custom post types (`ap_actor`, `ap_inbox`, `ap_outbox`, extra fields) are properly registered
4. Verify post meta fields are correctly associated with their respective post types
5. Test ActivityPub functionality to ensure no regressions

obenland committed:
Consolidate custom post type and meta registrations into Post_Types class

- Created new Post_Types class to manage all custom post type and meta registrations
- Moved post type registrations from Activitypub::register_post_types() to dedicated methods
- Consolidated post meta registration with each post type for better organization
- Removed duplicate post meta registrations from Blocks class
- Updated tests to use new Post_Types class methods
- Maintained user meta registrations in original Activitypub class location
- All 895 tests pass, ensuring backward compatibility

obenland commented on includes/class-post-types.php:
```diff
@@ -0,0 +1,434 @@
+			array(
+				'type'              => 'string',
+				'single'            => false,
+				'sanitize_callback' => 'esc_sql',
```
@pfefferle Should this be `absint`?

pfefferle replied:
technically it could be `-1` so `absint` might not be the best callback.

obenland replied:
`sanitize_text_field()` feels good then

pfefferle commented on includes/class-post-types.php:
```diff
@@ -0,0 +1,434 @@
+	 * Initialize the class, registering all custom post types and post meta.
+	 */
+	public static function init() {
+		self::register_remote_actors_post_type();
+		self::register_inbox_post_type();
+		self::register_outbox_post_type();
+		self::register_extra_fields_post_types();
+		self::register_activitypub_post_meta();
```
no alphabetic order? 😱

obenland replied:
I KNOW!!!

obenland commented on includes/class-activitypub.php:
```diff
@@ -949,7 +608,7 @@ public static function register_user_meta() {
 			array(
 				'type'              => 'string',
-				'description'       => 'The user’s description.',
+				'description'       => 'The user description.',
```
These changes are from Claude copying user meta registration back-and-forth. I left the changes in as these are "texturized" quotes, not translated, and the text still makes sense without them.
Happy to revert if you feel differently.

pfefferle commented on includes/class-blocks.php:
```diff
@@ -32,83 +31,6 @@ public static function init() {
 		\add_filter( 'activitypub_the_content', array( self::class, 'remove_post_transformation_callbacks' ) );
 	}
 
-	/**
```
Nice, I just wanted to bring this up!!

obenland commented:
@pfefferle Should `prevent_empty_post_meta()` and `default_post_metadata()` also be moved into `Post_Types`?

obenland committed:
Remove extra blank line in Activitypub class

Deleted an unnecessary blank line for improved code cleanliness in the Activitypub class.

obenland committed:
Fix regex pattern in coverage warning detection for escaped backslashes

The regex pattern was failing to match @covers warnings with escaped backslashes
in class names. Changed from .* to [^"]* to properly capture content within quotes.

obenland committed:
Simplify warning detection to just look for WARNINGS! marker

- Changed from trying to match specific warning count format to just looking for WARNINGS!
- Updated sed pattern to handle both singular and plural warning formats
- This is more robust and handles any number of warnings

pfefferle commented:
If we also move every meta registration I am not sure if Post_Type is still the right name!?

obenland commented:
> If we also move every meta registration I am not sure if Post_Type is still the right name!?

I guess I defined it generously in my head, anything post types related, including built-in post types.

obenland committed:
Fix invalid @covers annotations in tests

- Updated @covers annotation in test_activity_meta_sanitization to reference Post_Types::register_outbox_post_type
- Updated @covers annotation in test_register_post_meta to reference Post_Types::register_activitypub_post_meta
- These methods were moved from Activitypub::register_post_types and Blocks::register_postmeta to the new Post_Types class

obenland committed:
Deduplicate coverage warning annotations for dataProvider tests

- Added tracking of processed file:line combinations to avoid duplicate annotations
- Tests with dataProvider create multiple warnings for the same @covers annotation
- Now only creates one checkstyle annotation per unique file/line location
- Reduces noise in PR annotations while still catching all unique issues

pfefferle commented:
> Should `prevent_empty_post_meta()` and `default_post_metadata()` also be moved into `Post_Types`?

When we talk about these, we would also have to move:

* `prevent_empty_post_meta`
* `register_user_meta`

But then, the class would be very bloated again.

Should we only have the Post_Types defined there and introduce two other classes for `user_meta` and `post_meta`?

pfefferle commented:
I think I would move all the post_metas to the `Post_Types` class and we should think about having a `User_Metas` class.

obenland commented:
> When we talk about these, we would also have to move:
>
> * prevent_empty_post_meta
> * register_user_meta

What's the connection between `prevent_empty_post_meta` and `register_user_meta`?

pfefferle commented:
> What's the connection between prevent_empty_post_meta and register_user_meta?

There is none, except what I wrote above: We might want to also have a dedicated space for user metas. That said, I had a look at the code and the metas and options are all over the place, because it is not always the best solution to have them in one place (options for example). So I think we are fine with your PR for now.

obenland committed:
Move post meta logic to Post_Types and add tests

Refactored prevent_empty_post_meta and default_post_metadata methods from Activitypub to Post_Types for better separation of concerns. Updated hooks in Activitypub to use the new Post_Types methods. Added unit tests for these methods in a new Test_Post_Types class, and removed related tests from Test_Activitypub.

obenland committed:
Refactor Post_Types meta registration to use helper methods

Extract duplicate post meta args into reusable private helper methods:
- get_object_id_meta_args() for _activitypub_object_id
- get_activity_actor_meta_args() for _activitypub_activity_actor
- get_activity_type_meta_args() for _activitypub_activity_type
- get_content_visibility_meta_args() for activitypub_content_visibility

This eliminates code duplication between inbox/outbox registrations and ensures consistency across all post meta configurations.

obenland commented on includes/class-post-types.php:
```diff
@@ -0,0 +1,476 @@
+				$schema = array(
+					'type'    => 'string',
+					'enum'    => Activity::TYPES,
+					'default' => $default_type,
```
@pfefferle There's a difference in default activity type meta between Inbox and Outbox post types, where it's `Create` for Inbox and `Announce` for Outbox. Is that on purpose?

pfefferle replied:
We do not store `Announce`s in the inbox yet, so yes!

obenland committed:
Update sanitize_callback to use sanitize_text_field

Replaces custom and esc_sql sanitize_callback functions with WordPress's built-in sanitize_text_field for better input sanitization and consistency.

obenland committed:
Revert phpunit.yml changes to match trunk

Remove duplicate annotation tracking logic that was added in this branch.
Will be fixed in a separate PR focused on CI improvements.

obenland committed:
Merge branch 'trunk' into consolidate-post-type-registrations

pfefferle commented on includes/class-post-types.php:
```diff
@@ -119,61 +119,19 @@ public static function register_inbox_post_type() {
-					return $value;
-				},
-			)
+			self::get_activity_actor_meta_args()
```
I understand why you did this, but it makes readability a lot worse! In that case I would prefer to do it the same way we did is with Extra_Fields: https://github.com/Automattic/wordpress-activitypub/blob/dbf0d11631fa74ba00b56c3ed51746d582cb40b6/includes/class-activitypub.php#L776

obenland replied:
Reverted in [36ec110](https://github.com/Automattic/wordpress-activitypub/pull/2215/commits/36ec110d55f757e9aa833deb618dedbe1bbb22cd)

obenland committed:
Revert "Refactor Post_Types meta registration to use helper methods"

This reverts commit 1cd303fa7cf009dafee62950d8ba6ad9bbe1a8e5.

obenland committed:
Use same activitypub_content_visibility args everywhere

obenland committed:
Merge branch 'consolidate-post-type-registrations' of https://github.com/Automattic/wordpress-activitypub into consolidate-post-type-registrations

pfefferle commented on includes/class-activitypub.php:
```diff
@@ -45,9 +42,9 @@ public static function init() {
-		\add_filter( 'default_post_metadata', array( self::class, 'default_post_metadata' ), 10, 3 );
+		\add_filter( 'add_post_metadata', array( Post_Types::class, 'prevent_empty_post_meta' ), 10, 4 );
+		\add_filter( 'update_post_metadata', array( Post_Types::class, 'prevent_empty_post_meta' ), 10, 4 );
+		\add_filter( 'default_post_metadata', array( Post_Types::class, 'default_post_meta_data' ), 10, 3 );
```
why keep these here? I would prefer to move them too.

obenland replied:
oh shoot, that's right, we have a `init()` there

obenland committed:
Move post meta filters to Post_Types class

Relocates the registration of post meta-related filters from Activitypub to the Post_Types class for better organization and separation of concerns.

Props @pfefferle.

obenland committed:
Move register_ap_actor_rest_field to Post_Types class

The register_ap_actor_rest_field method and its related rest_api_init hook have been moved from the Activitypub class to the Post_Types class. This change centralizes REST field registration logic related to post types, improving code organization.

obenland committed:
Move ActivityPub post type and extra fields hooks

Relocates the registration of ActivityPub support for custom post types and the actor extra fields filter from class-activitypub.php to class-post-types.php. This change improves code organization by grouping post type related logic within the Post_Types class.

pfefferle commented on includes/class-post-types.php:
```diff
@@ -0,0 +1,548 @@
+
+		\add_action( 'rest_api_init', array( self::class, 'register_ap_actor_rest_field' ) );
+
+		\add_filter( 'activitypub_get_actor_extra_fields', array( Extra_Fields::class, 'default_actor_extra_fields' ), 10, 2 );
```
This feels a bit like a hack, we did because we wanted to keep the collections free from `init` hooks!

What do you think about also move the `default_actor_extra_fields` function to this class?

obenland replied:
As an alternative, how about we call `default_actor_extra_fields()` from `get_actor_fields()` when they've not been created yet? That would also avoid the side effect of creating posts in a filter callback. Happy to do that in a follow-up.

pfefferle replied:
I think we should keep the filter version, because not everyone might want to have the default fields. With the filter they can unhook the defaults and/or add their own ones.

obenland replied:
Is that really a common use case? Would you not expect them to just delete these fields in the UI?

pfefferle replied:
It is about site owners. Like https://blogs.vivaldi.net/ for example.

obenland replied:
I'd leave it as is then. Moving it would break any existing unhooks. It's fine to stay in that class.
